src/simulator/ackermann_model.cpp

```diff
@@ -0,0 +1,99 @@
+#include "simulator/ackermann_model.h"
+#include "shared/util/timer.h"
+#include "shared/math/math_util.h"
+
+using f1tenth_simulator::AckermannCurvatureDriveMsg;
+using math_util::AngleDiff;
+using math_util::AngleMod;
+using std::string;
+using std::isfinite;
+
+namespace ackermann {
+
+CONFIG_FLOAT(cMinTurnR, "min_turn_radius");
+CONFIG_FLOAT(cMaxAccel, "max_accel");
+CONFIG_FLOAT(cMaxSpeed, "max_speed");
+CONFIG_FLOAT(cAngularErrorBias, "angular_error_bias");
+CONFIG_FLOAT(cAngularErrorRate, "angular_error_rate");
+CONFIG_STRING(cDriveTopic, "drive_callback_topic");
+
+
+AckermannModel::AckermannModel(const string& config_file, ros::NodeHandle n) :
+    EntityBase(),
+    angular_error_(0, 1),
+    config_reader_({config_file}){
+  // Use the config reader to initialize the subscriber
+  drive_subscriber_ = n.subscribe(
+      cDriveTopic,
+      1,
+      &AckermannModel::DriveCallback,
+      this);
+}
+
+void AckermannModel::DriveCallback(const AckermannCurvatureDriveMsg& msg) {
+  if (!isfinite(msg.velocity) || !isfinite(msg.curvature)) {
+    printf("Ignoring non-finite drive values: %f %f\n",
+        msg.velocity,
+        msg.curvature);
+    return;
+  }
+  last_cmd_ = msg;
+  t_last_cmd_ = GetMonotonicTime();
+}
+
+float AbsBound(float x, float bound) {
+  if (x > 0.0 && x > bound) {
+    return bound;
+  } else if (x < 0.0 && x < -bound) {
+    return -bound;
+  }
+  return x;
+}
+
+void AckermannModel::Step(const double &dt) {
```

Comment:
No need to change this, but basic data types that fit the alignment of the computer see little to no benefits from passing by ref, since a pointer size is the same as the data itself. However, a ref places additional restrictions on the lifetime of the object from the caller. Better to just use `const double`.